src/bench/logging.cpp

```diff
@@ -17,32 +17,61 @@ static void Logging(benchmark::Bench& bench, const std::vector<const char*>& ext
     bench.run([&] { log(); });
 }
 
-static void LoggingYoThreadNames(benchmark::Bench& bench)
+static void LogPrintLevelWithThreadNames(benchmark::Bench& bench)
 {
-    Logging(bench, {"-logthreadnames=1"}, [] { LogPrintf("%s\n", "test"); });
+    Logging(bench, {"-logthreadnames=1"}, [] { LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Error, "%s\n", "test"); });
 }
-static void LoggingNoThreadNames(benchmark::Bench& bench)
+
+static void LogPrintLevelWithoutThreadNames(benchmark::Bench& bench)
 {
-    Logging(bench, {"-logthreadnames=0"}, [] { LogPrintf("%s\n", "test"); });
+    Logging(bench, {"-logthreadnames=0"}, [] { LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Error, "%s\n", "test"); });
+}
+
+static void LogPrintWithCategory(benchmark::Bench& bench)
+{
+    Logging(bench, {"-logthreadnames=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
+}
+
+static void LogPrintWithoutCategory(benchmark::Bench& bench)
```

Comment:
@kouloumos I'm seeing the same thing. After investigation, I found that the problem is that the global logging object instance is not re-created for each run. When `LogPrintWithoutCategory` runs, it still has logging categories enabled from the previous benchmark runs.

This patch fixes the problem for me, give it a try. The change to `LogInstance()` looks racy (i.e. what if two threads both see `g_logger` is null, and they both create a new object), but this change is safe because `LogInstance()` first runs as part of the [global static initialization](https://en.cppreference.com/w/cpp/language/initialization), before `main()` is called, so there's only one thread. But if there's concern about this, `std::call_once()` can be used (it's used in a few places already).

<details>
<summary>patch</summary>

```diff
diff --git a/src/bench/logging.cpp b/src/bench/logging.cpp
index d5c78932bc..434d1c58d4 100644
--- a/src/bench/logging.cpp
+++ b/src/bench/logging.cpp
@@ -9,6 +9,7 @@
 
 static void Logging(benchmark::Bench& bench, const std::vector<const char*>& extra_args, const std::function<void()>& log)
 {
+    LogReset();
     TestingSetup test_setup{
         CBaseChainParams::REGTEST,
         extra_args,
diff --git a/src/logging.cpp b/src/logging.cpp
index 8bf60169b0..24f3338368 100644
--- a/src/logging.cpp
+++ b/src/logging.cpp
@@ -18,6 +18,8 @@
 const char * const DEFAULT_DEBUGLOGFILE = "debug.log";
 constexpr auto MAX_USER_SETABLE_SEVERITY_LEVEL{BCLog::Level::Info};
 
+static BCLog::Logger* g_logger;
+
 BCLog::Logger& LogInstance()
 {
 /**
@@ -35,10 +37,16 @@ BCLog::Logger& LogInstance()
  * This method of initialization was originally introduced in
  * ee3374234c60aba2cc4c5cd5cac1c0aefc2d817c.
  */
-    static BCLog::Logger* g_logger{new BCLog::Logger()};
+    if (g_logger == nullptr) g_logger = new BCLog::Logger();
     return *g_logger;
 }
 
+void LogReset()
+{
+    delete g_logger;
+    g_logger = new BCLog::Logger();
+}
+
 bool fLogIPs = DEFAULT_LOGIPS;
 
 static int FileWriteStr(const std::string &str, FILE *fp)
diff --git a/src/logging.h b/src/logging.h
index 16f23a19e8..522923c000 100644
--- a/src/logging.h
+++ b/src/logging.h
@@ -266,6 +266,7 @@ namespace BCLog {
 } // namespace BCLog
 
 BCLog::Logger& LogInstance();
+void LogReset(); // testing only
 
 /** Return true if log accepts specified category, at the specified level. */
 static inline bool LogAcceptCategory(BCLog::LogFlags category, BCLog::Level level)
```
</details>